### PR TITLE
[core] Force kill worker whose job has exited

### PR DESCRIPTION
**Pull request by clarng**
## Why are these changes needed?

The worker leaks currently when the task references some global import like tensorflow. There are couple issues that led to this bug:
- when the worker finishes executing it does not clean up all its borrowed references
- the reference counting code treats borrowed reference as something it owns
- if the worker thinks it owns references it will not exit
- the worker pool will not force exit an idle worker, even if the job is dead, if the worker refuses to due to the aforementioned object ownership

This PR implements the logic in worker pool to force kill an idle worker whose job has exited

## Related issue number

https://github.com/ray-project/ray/issues/25699

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
> the reference counting code treats borrowed reference as something it owns


Hmm based on this description, isn't the right solution to distinguish borrowed objs from owned objs?

**Comment by clarng** [comment]
> > the reference counting code treats borrowed reference as something it owns
> 
> Hmm based on this description, isn't the right solution to distinguish borrowed objs from owned objs?

Fixing one of the 3 would fix this bug, but the worker pool's comment says it should kill worker whose job is dead (currently it does not), I think we should open separate issues for reference counting bug, I don't think there's bandwidth to fix both issues in the release timeframe as there are other blockers that needs to be fixed

**Comment by rkooo567** [comment]
Can you then open the reference counting bug?

File: src/ray/raylet/worker_pool.cc [hunk 1]
```diff
@@ -981,6 +982,9 @@ void WorkerPool::PushWorker(const std::shared_ptr<WorkerInterface> &worker) {
     int64_t now = get_time_();
     idle_of_all_languages_.emplace_back(worker, now);
     idle_of_all_languages_map_[worker] = now;
+  } else if (!found) {
```
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:985]: Have you actually seen this case? Should we create an issue?

File: src/ray/raylet/worker_pool.cc [hunk 2]
```diff
@@ -1100,6 +1120,7 @@ void WorkerPool::TryKillingIdleWorkers() {
               // In case of failed to send request, we remove it from pool as well
               // TODO (iycheng): We should handle the grpc failure in better way.
               if (!status.ok() || r.success()) {
+                RAY_LOG(DEBUG) << "Removed worker ";
```
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:1123]: Can you add worker id to all relevant logs below

**Review by rkooo567 — COMMENTED** [review]
Can you add a relevant unit test to worker_pool_test.cc?

**Comment by clarng** [comment]
> Can you then open the reference counting bug?

https://github.com/ray-project/ray/issues/32271

File: src/ray/raylet/worker_pool.cc (hunk 1 above)
Comment (clarng) [reply src/ray/raylet/worker_pool.cc:985]: https://github.com/ray-project/ray/issues/32271

Comment (scv119) [reply src/ray/raylet/worker_pool.cc:985]: nice catch!

**Review by scv119 — APPROVED** [review]

File: src/ray/raylet/worker_pool.cc [hunk 3]
```diff
@@ -1090,6 +1108,12 @@ void WorkerPool::TryKillingIdleWorkers() {
         RAY_CHECK(running_size > 0);
         running_size--;
         rpc::ExitRequest request;
+        if (finished_jobs_.contains(job_id) &&
+            RayConfig::instance().kill_idle_workers_of_terminated_job()) {
+          RAY_LOG(INFO) << "Force exiting worker whose job has exited"
+                        << worker->WorkerId();
+          request.set_force_exit(true);
```
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:1115]: Btw, it doesn't need to be done from this PR, but I think we should have more reliable way to force force kill workers than sending a RPC. E.g., send SIGKILL. For example, this issue https://ray-distributed.slack.com/archives/C03H0Q9CUQ3/p1675885877128899?thread_ts=1675460488.542369&cid=C03H0Q9CUQ3 won't be fixed. I will create an issue for this

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
lmk when tests pass

File: src/ray/raylet/worker_pool.cc (hunk 3 above)
Comment (clarng) [reply src/ray/raylet/worker_pool.cc:1115]: From the user's logs looks like they are calling multi-process from the worker, how do you propose we track those processes?

Comment (clarng) [reply src/ray/raylet/worker_pool.cc:1115]: Agree it is orthogonal to this PR

**Comment by clarng** [comment]
> lmk when tests pass

good to merge

had to fix another existing bug

**Review by rkooo567 — CHANGES_REQUESTED** [review]
Can you tell me what was the bug before merging?

**Comment by clarng** [comment]
> Can you tell me what was the bug before merging?

idle worker from alive job will break the loop and won't be able to kill idle worker from dead job ? 

https://github.com/ray-project/ray/pull/32217/commits/ecbe70c6debcf11f586435265281415bd4d91ff0

**Comment by cadedaniel** [comment]
ping for review @rkooo567

**Comment by rkooo567** [comment]
Oh that's a nice catch!

**Review by rkooo567 — APPROVED** [review]

**Comment by cadedaniel** [comment]
Let's merge!

**Comment by cadedaniel** [comment]
Sang is waiting for mac build to complete

**Comment by cadedaniel** [comment]
I think this failure https://buildkite.com/ray-project/oss-ci-build-pr/builds/11905#01863d1d-18af-4561-8ff0-2db28d41a301 is just flaky, rerunning.

**Comment by cadedaniel** [comment]
Mac wheels fails with `ValueError: When connecting to an existing cluster, num_cpus and num_gpus must not be provided.`. Is this caused by this PR? @clarng

**Pull request merged**